torch/csrc/jit/unpickler.cpp

```diff
@@ -621,12 +621,55 @@ PickleOpCode Unpickler::readInstruction() {
   return opcode;
 }
 
+void Unpickler::readSlowWithBuffer(char *dest, size_t sz) {
+  // First, read any partial from buffer (may be 0).
+  // We explicitly assume that sz > buffer_remaining_,
+  // and that sz is never bigger than buffer_.size().
+  AT_ASSERT(sz > buffer_remaining_);
+  const size_t from_old_buf = buffer_remaining_;
+  if (from_old_buf != 0) {
+    memcpy(dest, buffer_.data() + buffer_pos_, from_old_buf);
+  }
+  const size_t needed = sz - from_old_buf;
+  // Full read into the buffer. The calls here all explicitly
+  // assume that one buffer will be enough for any sz.
+  AT_ASSERT(sz <= buffer_.size());
+  buffer_remaining_ = reader_(buffer_.data(), buffer_.size());
+  if (buffer_remaining_ < needed) {
+    AT_ERROR("Unexpected end of pickler archive.");
+  }
+  memcpy(dest + from_old_buf, buffer_.data(), needed);
+  buffer_pos_ = needed;  // assignment (0'ed from read)
+  buffer_remaining_ -= needed;
+}
+
 // Read a number of bytes from the input stream
 std::string Unpickler::readBytes(size_t length) {
   std::string data(length, 0);
-  // This is fine since C++11 has contiguous strings
-  if (!reader_(&data[0], length)) {
-    AT_ERROR("Unexpected end of pickler archive.");
+  static const size_t kSmallString = 64;
+  if (length <= buffer_remaining_) {
+    // Fast-path: entirely in buffer.
+    memcpy(&data[0], buffer_.data() + buffer_pos_, length);
+    buffer_pos_ += length;
+    buffer_remaining_ -= length;
+  } else if (length <= kSmallString) {
+    // If the string is smallish, do a full buffer read,
+    // and read out of that buffer.
+    readSlowWithBuffer(&data[0], length);
+  } else {
+    // Otherwise, for larger strings, read what we can from
+    // the buffer, and then read directly to the destination.
+    const size_t from_old_buf = buffer_remaining_;
+    if (from_old_buf != 0) {
+      memcpy(&data[0], buffer_.data() + buffer_pos_, from_old_buf);
+    }
+    const size_t needed = length - from_old_buf;
+    size_t nread = reader_(&data[from_old_buf], needed);
```

Comment:
Shouldn't we do an over-read here?  Otherwise won't the buffer usually be empty?

Reply:
It this is because the `needed` might be larger than the buffer size (it is a string of any length), so it cannot read directly into the buffer because there might not be enough room there. There is a tradeoff in reading to the buffer and then having to copy to the string when the size is already big as well and the expected overhead of the reader_ call is relatively smaller.

Reply:
Right, the reason is the arbitrary length.
It seems that there is a threshold (maybe about a cache line or two?) where double-buffering cost is effectively small, but then once it grows too big, that reading into the end location can make more sense.